ni_measurementlink_service/measurement/service.py

```diff
@@ -183,10 +182,54 @@ def convert_value_to_str(value: object) -> str:
 
         self.configuration_parameter_list: List[Any] = []
         self.output_parameter_list: List[Any] = []
-        self.grpc_service = GrpcService()
-        self.context: MeasurementContext = MeasurementContext()
-        self.channel_pool: GrpcChannelPool = GrpcChannelPool()
-        self.discovery_client: DiscoveryClient = DiscoveryClient()
+        self.context = MeasurementContext()
+
+        self._initialization_lock = threading.RLock()
```

Comment:
I read a little about RLock vs Lock, but can you explain why the different choice in service and discovery_client?

Reply:
RLock is recursive. Lock isn't.

service has multiple lazily-initialized properties that call each other, so it needs a recursive lock.

discovery_client has a single lazily-initialized property, so it doesn't need a recursive lock.

Reply:
Ahh, thus the 'R'. :)

Reply:
Yeah, also not to be confused with RWLock (which is standard in some languages but not Python)